main/modes/quickSettings/quickSettings.c

```diff
@@ -0,0 +1,337 @@
+    .fnExitMode               = quickSettingsExitMode,
+    .fnMainLoop               = quickSettingsMainLoop,
+    .fnAudioCallback          = NULL,
+    .fnBackgroundDrawCallback = NULL,
+    .fnEspNowRecvCb           = NULL,
+    .fnEspNowSendCb           = NULL,
+    .fnAdvancedUSB            = NULL,
+};
+
+/// All state information for the Quick Settings mode. This whole struct is calloc()'d and free()'d
+/// so that Quick Settings is only using memory while it is active
+quickSettingsMenu_t* quickSettings = NULL;
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+/**
+ * @brief Helper function to easily setup the quick settings options values.
+ *
+ * @param bounds The setting bounds to use, with min for the "off" value
+ * @param currentValue The current setting value, to be used as the "on" value if larger than the min
+ * @param optionValues Pointer to an array of 2 values where the values will be written
+ * @return int32_t Returns the current setting value for convenience
+ */
+static int32_t setupQuickSettingParams(const settingParam_t* bounds, int32_t currentValue, int32_t* optionValues)
+{
+    // Use the minimum and the current value for the toggles
+    // This way, we get the current value back after toggling off.
+    optionValues[0] = bounds->min;
+    optionValues[1] = currentValue;
+
+    // However, if it was already off / at the minimum, we need it to be something, so just use the max
+    if (currentValue <= bounds->min)
+    {
+        optionValues[1] = bounds->max;
+    }
+
+    return currentValue;
+}
+
+/**
+ * @brief Enter Pong mode, allocate required memory, and initialize required variables
+ *
+ */
+static void quickSettingsEnterMode(void)
+{
+    // Allocate and clear all memory for this mode. All the variables are contained in a single struct for convenience.
+    // calloc() is used instead of malloc() because calloc() also initializes the allocated memory to zeros.
+    quickSettings = calloc(1, sizeof(quickSettingsMenu_t));
+
+    // Load a font
+    loadFont("ibm_vga8.font", &quickSettings->font, true);
+
+    // Load graphics
+    // Use SPI because we're not the only mode, I guess?
+    loadWsg("defaultSetting.wsg", &quickSettings->iconGeneric, true);
+    loadWsg("ledsEnabled.wsg", &quickSettings->iconLedsOn, true);
+    loadWsg("ledsDisabled.wsg", &quickSettings->iconLedsOff, true);
+    loadWsg("musicEnabled.wsg", &quickSettings->iconBgmOn, true);
+    loadWsg("musicDisabled.wsg", &quickSettings->iconBgmOff, true);
+    loadWsg("sfxEnabled.wsg", &quickSettings->iconSfxOn, true);
+    loadWsg("sfxDisabled.wsg", &quickSettings->iconSfxOff, true);
+
+    // Initialize the menu
+    quickSettings->menu     = initMenu(quickSettingsName, quickSettingsMenuCb);
+    quickSettings->renderer = initMenuQuickSettingsRenderer(&quickSettings->font);
```

Comment:
I'm so happy that there is a second renderer. It really validates splitting the data structure from the UI 😊

Reply:
What can I say, I'm a sucker for a well defined interface!